### DISCUSSION TITLE
Drift detected on first migration after using migrate diff on existing database

### DISCUSSION THREAD

**Discussion by MikhailKochkin** [prisma/prisma#20025] [Q&A] [answered]
### Bug description

I am attempting to add Prisma 4.16 to an existing project (it used to use Prisma 2.12 ) on a MacBook Pro with an M1 Pro chip. During the upgrade process, I am encountering an issue with Prisma's migration system. Despite following the upgrade documentation (https://www.prisma.io/docs/getting-started/setup-prisma/add-to-existing-project/relational-databases/connect-your-database-typescript-postgresql) and baselining the database as per the guidelines, I am encountering a drift error when trying to migrate.

The error message states: "Drift detected: Your database schema is not in sync with your migration history." The error message also provides a summary of the differences between the expected database schema and the actual schema of the database, which includes changes to several tables and foreign keys.

```
[+] Added tables
  - _Migration

[*] Changed the `CareerTrack` table
  [-] Removed column `type`

[*] Changed the `_AuthorsCoursePage` table
  [-] Removed foreign key on columns (A)
  [-] Removed foreign key on columns (B)
  [+] Added foreign key on columns (A)
  [+] Added foreign key on columns (B)

[*] Changed the `_CareerTrackUnitToCoursePage` table
  [-] Removed foreign key on columns (A)
  [-] Removed foreign key on columns (B)
  [+] Added foreign key on columns (A)
  [+] Added foreign key on columns (B)

[*] Changed the `_CoursePageForCareer` table
  [-] Removed foreign key on columns (A)
  [-] Removed foreign key on columns (B)
  [+] Added foreign key on columns (A)
  [+] Added foreign key on columns (B)

[*] Changed the `_CoursePageToProgram` table
  [-] Removed foreign key on columns (A)
  [*] Altered column `A` (type changed)
  [+] Added foreign key on columns (A)

[*] Changed the `_ExamAnswerToLegalPortfolio` table
  [-] Removed foreign key on columns (A)
  [-] Removed foreign key on columns (B)
  [*] Altered column `A` (type changed)
  [*] Altered column `B` (type changed)
  [+] Added foreign key on columns (A)
  [+] Added foreign key on columns (B)

[*] Changed the `_GrowthAreaToUserLevel` table
  [-] Removed foreign key on columns (B)
  [*] Altered column `B` (type changed)
  [+] Added foreign key on columns (B)

[*] Changed the `_PointATests` table
  [-] Removed foreign key on columns (A)
  [-] Removed foreign key on columns (B)
  [*] Altered column `A` (type changed)
  [*] Altered column `B` (type changed)
  [+] Added foreign key on columns (A)
  [+] Added foreign key on columns (B)

[*] Changed the `_UserSubjects` table
  [-] Removed foreign key on columns (A)
  [-] Removed foreign key on columns (B)
  [+] Added foreign key on columns (A)
  [+] Added foreign key on columns (B)

[*] Changed the `_joinedTeams` table
  [-] Removed foreign key on columns (B)
  [*] Altered column `B` (type changed)
  [+] Added foreign key on columns (B)
```

Despite my attempts to resolve the issue, including creating a new migration, modifying the Prisma schema, and manually editing the migration file, the drift error persists. Any command involving prisma migrate dev results in a drift error.

It's important to note that the database tables are fine, no manual changes have been made to the, and contain important data that cannot be lost. The tables that cause the problem are part of implicit many-to-many relations that are not specifically defined in the Prisma schema.

I have found related discussions on GitHub ([#17558](https://github.com/prisma/prisma/issues/17558), [#11555](https://github.com/prisma/prisma/issues/11555)), but they cannot help in my situation.

### How to reproduce

I can give access to the db uri in private.

All other files were taken from the guideline on the website: https://www.prisma.io/docs/getting-started/setup-prisma/add-to-existing-project/relational-databases/connect-your-database-typescript-postgresql

### Expected behavior

Npx migrate dev should apply changes to the DB

### Prisma information

```prisma
generator client {
  provider = "prisma-client-js"
}

datasource db {
  provider          = "postgresql"
  url               = env("DATABASE_URL")
  shadowDatabaseUrl = env("SHADOW_DATABASE_URL")
}

model Application {
  id            String     @id @db.VarChar(30)
  applicantId   String     @db.VarChar(25)
  applicantName String
  message       String?
  coursePageID  String     @db.VarChar(25)
  updatedAt     DateTime
  createdAt     DateTime   @default(now())
  coursePageId  String     @db.VarChar(30)
  promocode     String?
  coursePage    CoursePage @relation(fields: [coursePageId], references: [id], onDelete: NoAction, onUpdate: NoAction)
}

model BotDialogue {
  id        String   @id
  journey   String[]
  createdAt DateTime @default(now())
  updatedAt DateTime
  rating    Int?
  source    String?
}

....

```

### Environment & setup

- OS: macOS
- Database: PostgreSQL
- Node.js version: v14.21.3 / 16.14

### Prisma Version
```
prisma                  : 4.16.2
@prisma/client          : 4.16.2
Current platform        : darwin-arm64
Query Engine (Node-API) : libquery-engine 4bc8b6e1b66cb932731fb1bdbbc550d1e010de81 (at node_modules/@prisma/engines/libquery_engine-darwin-arm64.dylib.node)
Migration Engine        : migration-engine-cli 4bc8b6e1b66cb932731fb1bdbbc550d1e010de81 (at node_modules/@prisma/engines/migration-engine-darwin-arm64)
Format Wasm             : @prisma/prisma-fmt-wasm 4.16.1-1.4bc8b6e1b66cb932731fb1bdbbc550d1e010de81
Default Engines Hash    : 4bc8b6e1b66cb932731fb1bdbbc550d1e010de81
Studio                  : 0.484.0
```

**Comment by Jolg42** [comment]
@MikhailKochkin `2.12` is quite old now, it's a very good idea to upgrade! Though I recommend you to first read these upgrade guides:
- https://www.prisma.io/docs/guides/upgrade-guides/upgrading-versions/upgrading-to-prisma-3
	- and especially https://www.prisma.io/docs/guides/upgrade-guides/upgrading-versions/upgrading-to-prisma-3/referential-actions
- https://www.prisma.io/docs/guides/upgrade-guides/upgrading-versions/upgrading-to-prisma-4

I'm not sure what would be the best course of action, though I'm thinking that first upgrading to the latest v3 and then latest v4 might be a good idea.

(Sidenote: Prisma v5 is coming very soon, but should be a lot easier to upgrade than the previous ones!)

**Reply by MikhailKochkin** [reply]
@Jolg42 Thank you. I have downgraded to Prisma@3.15.0

I have used guidelines on the Prisma website. I:

1. Looked through the Referential actions upgrade path. Introspected the DB. Fixed all the cascade cases.
2. Looke into name constraints. Introspected the DB. Found no naming problems, no map attributes were added to the Schema.

Started from scratch: deleted migrations folder in the project, deleted my initial migration from DB in _prisma_migrations table.

Then went through:

– mkdir -p prisma/migrations/0_init
– npx prisma migrate diff --from-empty --to-schema-datamodel prisma/schema.prisma --script > prisma/migrations/0_init/migration.sql
– npx prisma migrate resolve --applied 0_init => Migration 0_init marked as applied.
– npx prisma generate 

Successfully queried the DB

Then I make a little change to the Schema. Run npx prisma migrate dev --name tags-model. And again get:
Drift detected: Your database schema is not in sync with your migration history.

**Reply by jason9717** [reply]
You are a life saver bro. Thank u so much)

**Comment by hajnalben** [comment]
I am experiencing the same issue that @MikhailKochkin described. Currently migrating my prisma1 schema to v4. Everything looks perfect. 
Created an initial migration file by:
`npx prisma migrate diff --from-empty --to-schema-datamodel prisma/schema.prisma --script > prisma/migrations/0_init/migration.sql`
Then applied this migration by:
`npx prisma migrate resolve --applied 0_init`
I can see it in the _prisma_migrations table

However when I try to add a new field to any of my models and run `prisma migrate dev` the same Drift detected message appears and wants to reset my data in the db.

**Comment by hajnalben** [comment]
After some investigation:
 - applying the `prisma migrate dev` to a brand new DB and comparing it with my updated v1 DB I have found out that there are differences because of the changes in default referential actions: https://www.prisma.io/docs/guides/upgrade-guides/upgrading-versions/upgrading-to-prisma-3/referential-actions

**Comment by MikhailKochkin** [comment]
I applied `prisma migrate dev` to my shadow DB which is empty.

It again showed the error drift message and suggested resetting the DB. I resetted it successfully. I got:

```
Applying migration `0_init`

The following migration(s) have been applied:

migrations/
  └─ 0_init/
    └─ migration.sql
Applying migration `20230630172142_tags_model`

The following migration(s) have been created and applied from new schema changes:

migrations/
  └─ 20230630172142_tags_model/
    └─ migration.sql
    
   Your database is now in sync with your schema.

✔ Generated Prisma Client (3.15.0 | library) to ./node_modules/@prisma/client in 741ms

```

**I checked out the schema. It removed @relation attributes on my many-to-many connections..**

coursePage    CoursePage[] @relation("CareerTrackUnitToCoursePage") => coursePage    CoursePage[]

**But when I used this updated schema on my main DB with data I again got a drift error with the same comments. 🙁**

**Comment by vasujogani** [comment]
Tagging is here for visibility - https://github.com/prisma/prisma/discussions/20368 (Drfit detected error)

**Comment by MikhailKochkin** [comment, marked as answer]
It seems I've found the solution.

The error was with the migration.sql file which was in conflict with the actual database schema.

So I decided to look at how the file is generated.

We use `migrate diff` command to do that. According to the [Prisma guide](https://www.prisma.io/docs/guides/migrate/developing-with-prisma-migrate/add-prisma-migrate-to-a-project), we should run the following command:

```
npx prisma migrate diff \ 
--from-empty \ 
--to-schema-datamodel prisma/schema.prisma \ 
--script > prisma/migrations/0_init/migration.sql
```

So it uses the datamodel to generate migration.sql. I decided to change the datamodel to datasource (the db) to generate a new migration.sql. Which seemed like a reasonable idea since I had a conflict with the DB schema itself.

And it worked. The system no longer asked me to reset anything and only made some insignificant changes to the many-to-many relation tables. And now I can apply migration to the databases.

I hope my experience can help others facing similar issues.

By the way, I upgraded to Prisma 5 before implementing these changes.

**Reply by razvan-soare** [reply]
Hey @MikhailKochkin could you please explain a bit what do you mean by "change the datamodel to datasource (the db) to generate a new migration.sql." I am facing the same exact problem but still have no idea how to deal with it. 

Whenever i do a new migration is detecting some drift and not sure if im supposed to fix it in the initial schema or generate it in a diferent way.  Im using Prisma 5 too. Thank you

**Reply by MikhailKochkin** [reply]
Hello @razvan-soare, 

I mean that the docs tell you to run the following command to generate the migration.sql file:

```
npx prisma migrate diff \ 
--from-empty \ 
--to-schema-datamodel prisma/schema.prisma \ 
--script > prisma/migrations/0_init/migration.sql
```

This command uses the datamodel to generate the file. But the datamodel has some inherent flaws that cause the problem.
So try generating the migration.sql file based on the db itself instead.

To do that, update the command to use the datasource instead of datamodel. I don't remember the exact command, probably something like this:

```
npx prisma migrate diff \ 
--from-empty \ 
--to-schema-datasource prisma/schema.prisma \ 
--script > prisma/migrations/0_init/migration.sql
```

You can read more about different ways to generate initial migration and the difference between datamodel / datasource [**here**](https://www.prisma.io/docs/guides/migrate/developing-with-prisma-migrate/generating-down-migrations).

Hope that helps.
